git/gogit/client.go

```diff
@@ -251,6 +251,49 @@ func (g *Client) Clone(ctx context.Context, url string, cfg repository.CloneConf
 		return nil, err
 	}
 
+	var (
+		providerCreds *git.Credentials
+		expiresOn     time.Time
+		err           error
+	)
+
+	if g.authOpts != nil && g.authOpts.ProviderOpts != nil && g.authOpts.BearerToken == "" {
+		log := logr.FromContextOrDiscard(ctx)
+		// If cache is specified, try cloning with cached credential if it exists,
+		// else fallback to fetching credentials from provider.
+		if g.authOpts.Cache != nil {
+			cachedCreds, exists, err := getObjectFromCache(g.authOpts.Cache, url)
```

Comment:
What if there are hundreds of DevOps repos in the same Azure account, shouldn't we cache the token based on the Azure account instead of using the repo URL?

Reply:
Ok we can add the UID to the client options and use it as the cache key, while this will avoid conflicts it doesn't make Azure DevOps OIDC fit for multi-tenancy as there is no way to restrict a tenant from accessing another tenant repos.

Reply:
> In this per object identity case, I think we can't simplify anything and have to use the full repo URL as cache key.

Heavy plus one on this, as different identities may have different permissions to the same repository.
An additional point of caution here is cache invalidation when the object identity is removed/changed, which makes me think that potentially concatenating both identity metadata plus the UID may be a sensible approach.

Reply:
Thanks for the discussions. 

The reason I had kept the URL as the cache key was because of the way fine-grained permissions can be assigned at a per-project/per-repository level in Azure DevOps. From source controller's perspective, since there is a single (workload) identity that can be setup, in a happy path, this identity would need to be given access to all repositories that have the `provider` field in `GitRepository` spec set to `azure` to work correctly. The token associated with this identity would be fetched and cached once per repository. If the user sets things up incorrectly, for example: where there are multiple repositories and the identity is not given access to one of the repositories that has `provider` set to `azure`, then only the incorrectly configured repository fails with authentication error (as it should). The cached token is still valid for the repositories where permissions are setup correctly. 

Using `GitRepository` object's UID instead of repo URL should behave similarly and can be used as a cache key as well if this makes the approach common for other providers (Github, Gitlab). I can make the changes needed for this. 

Regarding cache invalidation, the approach is to treat the cache as best-effort. If there are any errors encountered while using the cached credentials, we fall back to fetching and using the credentials from the provider. If there are errors encountered while using the cached credential or while interacting with the cache (set/get), these are logged for debugging purposes but not propagated as user-facing errors.

Reply:
> different identities may have different permissions to the same repository

@pjbgf how can you have different identities with Workload ID?